pkg/cloudprovider/cloudapi/azure/azure_security.go

```diff
@@ -297,21 +297,65 @@ func (computeCfg *computeServiceConfig) processAsgAttachDetachConcurrently(asgOb
 	return err
 }
 
+// splitCloudRulesByDirection splits the given securitygroup.CloudRule slice into two, one for ingress rules and one for egress rules.
+func splitCloudRulesByDirection(rules []*securitygroup.CloudRule) ([]*securitygroup.CloudRule, []*securitygroup.CloudRule) {
+	ingressRules := make([]*securitygroup.CloudRule, 0)
+	egressRules := make([]*securitygroup.CloudRule, 0)
+	for _, rule := range rules {
+		switch rule.Rule.(type) {
+		case *securitygroup.IngressRule:
+			ingressRules = append(ingressRules, rule)
+		case *securitygroup.EgressRule:
+			egressRules = append(egressRules, rule)
+		}
+	}
+	return ingressRules, egressRules
+}
+
 // buildEffectiveNSGSecurityRulesToApply prepares the update rule cloud api payload from internal rules.
 func (computeCfg *computeServiceConfig) buildEffectiveNSGSecurityRulesToApply(appliedToGroupID *securitygroup.CloudResourceID,
-	ingressRules []*securitygroup.CloudRule, egressRules []*securitygroup.CloudRule, perVnetAppliedToNsgName string,
+	addRules []*securitygroup.CloudRule, rmRules []*securitygroup.CloudRule, perVnetAppliedToNsgName string,
 	rgName string) ([]*armnetwork.SecurityRule, error) {
 	// get current rules for applied to SG azure NSG
 	nsgObj, err := computeCfg.nsgAPIClient.get(context.Background(), rgName, perVnetAppliedToNsgName, "")
 	if err != nil {
 		return []*armnetwork.SecurityRule{}, err
 	}
-
 	if nsgObj.Properties == nil {
 		return nil, fmt.Errorf("properties field empty in nsg object %s", *nsgObj.ID)
 	}
-	var currentNsgIngressRules []armnetwork.SecurityRule
-	var currentNsgEgressRules []armnetwork.SecurityRule
+
+	addIRule, addERule := splitCloudRulesByDirection(addRules)
+	rmIRule, rmERule := splitCloudRulesByDirection(rmRules)
+
+	agAsgMapByNepheControllerName, atAsgMapByNepheControllerName, err := getNepheControllerCreatedAsgByNameForResourceGroup(
+		computeCfg.asgAPIClient, rgName)
+	if err != nil {
+		return []*armnetwork.SecurityRule{}, err
+	}
+	addIngressSecurityRules, err := convertIngressToNsgSecurityRules(appliedToGroupID, addIRule,
```

Comment:
can you convert one function to get all the rules, and do error comparison just once.

Reply:
I tried but I don't find a good way of doing it. Moving all these split and converts into a helper func works but the func signature is way too big and not readable. The convert function themselves are already complicated enough and I don't want to add more things into it. I shortened some names and reduced some lines here to help with cleaness.